### PR TITLE
rec: Fix DNSSEC validation with forward-zone-recurse

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Related to #7234.

Only did a few quick tests, needs a lot more testing and regression tests.
We also need to investigate the:

```
!(wasForwardedOrAuthZone && ne->d_auth != authname)) { // Only the authname nameserver can neg cache entries
```

removal because as of now this code makes no sense to me.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -2143,7 +2142,7 @@ RCode::rcodes_ SyncRes::updateCacheFromRecords(unsigned int depth, LWResult& lwr
     if (shouldValidate() && recordState == Secure) {
       vState initialState = recordState;
 
-      if (isAA) {
+      if (isAA || wasForwardRecurse) {
```
Comment (rgacogne) [inline pdns/syncres.cc]: This most likely breaks for `CNAME` answers containing more than the exact name sought.

**Comment by rgacogne** [comment]
Rebased on master because the DNSPython / DNSCrypt issue has been fixed there.

**Comment by pieterlexis** [comment]
>  We also need to investigate the:
> `!(wasForwardedOrAuthZone && ne->d_auth != authname)) { // Only the authname nameserver can neg cache entries`
> removal because as of now this code makes no sense to me.

Commit 64a8b6a1c1e1f165c8eb716f399e4da6d71156e7 has the reason:

> Make the negcache forwarded zones aware
>     
> Because of DNSSEC (without NTAs configured), we 'leak' the forwarded
> name to the cache when looking for the DNSKEY. This resulted in NXDOMAIN
> answers for actual NXDOMAINs, but with the root SOA record. Leading to
> inconsistent answers to the client.

We'll need to check what happens with this commit (it looks like no regression tests exist)

**Comment by pieterlexis** [comment]
I lied, #4377 adds tests for this issue (#4369)! So if travis says "OK", it hopefully is. Will do the code review now

**Comment by rgacogne** [comment]
> Commit 64a8b6a has the reason:

Yes, but that only makes sense to me for the first part of the commit, along `// when forwarding, the root may only neg-cache if it was forwarded to`. I don't get the second part, the one I removed in this PR.

**Comment by pieterlexis** [comment]
The second part existed for non-delegated subdomains not from the root afaik (but this was some time ago though, so I don't remember exactly)

**Pull request merged**